### ISSUE TITLE
[HELP]: Test setup

### ISSUE THREAD

**Issue by kentcdodds** [kentcdodds/testing-workshop#4]
Hey friends! I could use some help getting the setup for this thing tested! If you could please follow the instructions in the README and let me know:

1. Are they clear?
2. Does it work? (If not, please file an issue for the issue you're seeing)
3. What system are you on?
4. How could it be improved?

Thanks!

**Comment by callummr** [comment]
Windows 10, x64

First run I accidentally hadn't installed Mongo (my bad, readme is clear). It threw the relevant error (below) but then added `Congrats! Your system is setup properly. You should be good to install and run things.`

```
> testing-workshop@1.0.0 setup D:\dev\testing-workshop
> node ./scripts/verify.js && node ./scripts/install && npm start validate

The system cannot find the path specified.
there was an error determining your mongo version
{ Error: Command failed: "C:/Program Files/MongoDb/Server/3.2/bin/mongod.exe" --version
The system cannot find the path specified.
```

Not really a problem per se.

I then installed the latest mongo version for Windows (3.4.2 which is supported per readme) and got the same error. I noticed at line 74 of `verify.js` the version is hardcoded to 3.2

That's all for now as yarn is running incredibly slowly (problem on my end I'm pretty sure), will update with any more findings.

**Comment by kentcdodds** [comment]
That's really helpful! I fixed the congratulations thing. I'm not sure what I can do about the mongod version in the path on windows thing... Do you know how I could ensure that we're pointing to the right binary? I'd love to not have to have the full path...

**Comment by callummr** [comment]
Continued that convo in the issue I raised for it :)

Another thing, I set my git up with `core.autocrlf true` (ie. convert line endings to Windows on pull, back to unix on push) which I believe is pretty standard on Windows. Linting is pretty upset by this, outputting an error for every line of every file:
```
[lint] D:\dev\testing-workshop\api\models\User.js
[lint]     1:33  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
```

Not sure if the correct solution is changing the linting rules or git configuration. Most Windows editors will have no problem with LF line endings but newer developers or those only familiar with Windows might run in to issues later if they do change the git config.

**Comment by kentcdodds** [comment]
Ah, yeah, I keep forgetting about that line break stuff! I've added a `.gitattributes` file. You may need to re-clone to get that updated. Thanks!

**Comment by kentcdodds** [comment]
Hey @callummr, did the `npm start` script work for you? Did it pop open all the commands in different windows for you?

**Comment by edm00se** [comment]
Steps were clear, things didn't work for me in either of my main environments; details below.

<details><summary>### Windows 7 - Result: 💩☹️</summary>
This one rambles a bit, sorry about that.

- Git BASH
- [mongo installed from chocolatey](https://chocolatey.org/packages/mongodb) is at v3.2, making it in a different relative path to the hard-coded one in `package-scripts.j`, as @callummr mentioned (your req's do state >=3.4.2, more just an FYI I had to upgrade outside my usual Windows package manager)
  - installing the [latest (3.4.2) from mongodb.com](https://www.mongodb.com/download-center?jmp=nav#enterprise) actually parked things into the `C:\Program Files\MongoDB\Server\3.4` path
- no `yarn` in this environment, used `node ./scripts/install` as was mentioned in the output to the log from `npm run setup`
  - of note, the default small window of Git BASH means user has to scroll up a ways to see what the `npm` equivalent is (otherwise they're just staring at a lot of "npm ERR! ..."); this is the message from `scripts/verify.js`
  - install via `npm` took a little while, there looks to be a lot of goodies in there!
- `npm start` launches three `cmd.exe` shell windows, at the correct repo path, but didn't execute anything
  - got the same results from `cmd.exe` as opposed to Git BASH
  - the above was on node 6.9.1, npm 3.10, upgraded to 6.9.5 and npm@4(.2.0)
  - Git BASH spawned shells, again nothing running (same for `cmd.exe`)
- I created a symlink of 3.4.2 to 3.4
- ran again, from `cmd.exe` while running "as Administrator" and again poops out
</details>
<br />
<details><summary>### macOS Sierra (10.12.3) - Result: 🤔☹️</summary>

- node@6.9.5, npm@4.2.0
- w/ yarn@latest (0.20.3)
- mongodb@latest via homebrew (3.2.6, running, via `brew services ...`)
- [hyper](https://hyper.is/)
- install (`npm run setup`) was good, `./scripts/verify` succeeds
- `npm start` launches the three child shell instances, w/ nothing running in them
</details>

If I get some time tonight, I may take another pass at it.

**Comment by kentcdodds** [comment]
> hard-coded one in `package-scripts.js`

That's been dealt with. The instructions now ask you to add `mongod` to your PATH (with links showing you how to do that).

> otherwise they're just staring at a lot of "npm ERR! ..."

Changed the script to be `npm run setup --silent` so you don't see `npm ERR!` everywhere. Thanks!

> `npm start` didn't work

:-( If you have any tips on how to make this work, [here's how that happens](https://github.com/kentcdodds/testing-workshop/blob/master/package-scripts.js#L125-L135)...

Based on your experience with `npm start` on mac with hyper/zsh, I'm thinking maybe it's not worth putting too much effort into making it work. Maybe I should just tell people to run each script individually. Honestly, most of the time they wont be running the app anyway (except for the E2E tests). They'll just be running the test scripts.

**Comment by edm00se** [comment]
Interesting. The troubling part to me was that `terminal.app` w/ zsh works fine (in a directory path without a space), so it seems to be an issue directly w/ how things are being handled in hyper's new shell instance.

**Comment by kentcdodds** [comment]
Well, if you look at [the function](https://github.com/kentcdodds/testing-workshop/blob/master/package-scripts.js#L125-L135) that generates the script to run, it's pretty `Terminal.app` specific 😅 If we could make it work regardless of what terminal you use, we'd probably cover 80% of use cases.

**Comment by edm00se** [comment]
That makes perfect sense now. 🤔

**Comment by callummr** [comment]
@kentcdodds Sorry, early night. Just tried start and it launches three `cmd` instances but they aren't running anything. Think I see the issue, will PR if I can before I leave this morning.

**Comment by kentcdodds** [comment]
@edm00se, if you get a chance to give it a go again, please do. Thanks to @callummr's changes I think we're ready!

**Comment by edm00se** [comment]
That definitely fixes the Windows issue I noted with the spawned shells lacking scripts to run.

**Comment by kentcdodds** [comment]
I'm now struggling with this:

![emfile](https://cloud.githubusercontent.com/assets/1500684/23175389/7c2c3ef8-f81c-11e6-86c6-ac38c413de03.png)

![emfile2](https://cloud.githubusercontent.com/assets/1500684/23175390/7c2de730-f81c-11e6-8e82-d0680e14fdfd.png)

![emfile3](https://cloud.githubusercontent.com/assets/1500684/23175391/7c2ec150-f81c-11e6-845e-980f7e53962f.png)

It's more than annoying! I'm not sure what's up... I've done just about everything I can think of. I've added a [`TROUBLESHOOTING.md`](https://github.com/kentcdodds/testing-workshop/blob/master/other/TROUBLESHOOTING.md) in the hopes that if someone runs into this one of those things will help them...

Could I get some more people to try the setup? It's changed a bit and I want to see if anyone else gets this...

**Comment by paulfalgout** [comment]
While looking into your cypress setup I had no trouble with the initial install or getting the tests running.

The hard part was that I was unfamiliar with `nps`.  At first it seemed like a nice way to store the scripts and look up options available (and it looks very nice), but I incorrectly assumed I could easily scan for the options available.. Since `nps` was merely a pro tip, I didn't worry about installing it right away, but the `getE2EScripts` really obscured the available options, and so I had assumed at first that this lib _only_ ran Cypress headless.  While looking deeper at the e2e-cypress script I lucked out and saw the message about being "good to go" when devMode was true, but it wasn't clear that I needed to run `npm start e2e.dev` to get it in devMode.

All that to say, maybe a list of the most likely used tasks, or possibly a mention of `npm start help`.

Also a note that I never got the incoming connection message about Cypress...

**Comment by kentcdodds** [comment]
Hi @paulfalgout, thanks for the feedback!

I need to document the relevant scripts still. Most of the scripts are not useful by themselves and should probably be hidden from `npm start help`. I'll definitely do this. For what it's worth, this workshop is intended to be given with instruction, so hopefully people wont be lost with me right there to direct them and tell them what scripts to run. But it'd be great to document these things as well.

Also, as powerful as `nps` is, the available scripts can definitely be obscured by doing the logic that I'm doing in `getE2EScripts`. Makes _my_ life (script author) easier, but not the _users_ life (script user) easier.

> Also a note that I never got the incoming connection message about Cypress...

Could you elaborate on that?

**Comment by paulfalgout** [comment]
Sounds good.

The dialog popup in the screenshot under [Setup](https://github.com/kentcdodds/testing-workshop#setup)  For me, I don't get it running e2e..  Not sure if there's some way I could help determine why I never get it.. but I never did.

**Comment by kentcdodds** [comment]
Oh! That's great. I hope it's just a weird thing about my machine. I'm glad you didn't see it :D

**Comment by jennifer-mann** [comment]
This is nitpicking, but It would be helpful if the `verify.js` script could warn about the `node` version being out of date first, since `yarn` and `npm` are installed per `node` version. If someone were to update `yarn` / `npm` before seeing the `node` version warning, then they have to go back and re-install.

---

Also, I did get the Cypress tests to run successfully headlessly and I get that there is a lot of script setup to get that going, but I'm wondering if there are plans to have a process or `npm` script to run to get the Cypress tests to run within the Cypress app locally? Right now all of the tests fail due to the request urls being undefined.

**When running Cypress tests locally**
<img width="445" alt="screen shot 2017-03-13 at 11 13 00 am" src="https://cloud.githubusercontent.com/assets/1271364/23860833/715f76d0-07de-11e7-8279-3107087ded50.png">

**Comment by kentcdodds** [comment]
Great feedback.

For the second thing, run: `npm start e2e.dev`

Let me know whether that works!

**Comment by jennifer-mann** [comment]
Thanks, that did work! 

Unfortunately, [Cypress is not supported in Windows](https://github.com/cypress-io/cypress/issues/74) quite yet. I think this should probably be noted in the Readme and also maybe taken into account in the scripts so that the cypress tests do not run if on Windows. It doesn't necessarily *break* anything atm though and the process exits with code 0. The following prints in the console:

```bash
[cypress] '.' is not recognized as an internal or external command,
[cypress] operable program or batch file.
[cypress] shutting down cypress
```

**Comment by jennifer-mann** [comment]
Another recommendation. If I happen run `npm start e2e.dev` this opens a new process of Cypress for each time that it is run, so if I run `npm start e2e.dev` and `command + c` kill it a few times, I end up with multiple Cypress processes still running (and sitting in my dock). 

You may want to run `killall Cypress` on exiting to clean up the processes.

<img width="187" alt="screen shot 2017-03-13 at 11 58 57 am" src="https://cloud.githubusercontent.com/assets/1271364/23865618/a37f0e2e-07ec-11e7-94cf-33ecd9c27421.png">

**Comment by kentcdodds** [comment]
Whoa, just noticed this:

> Unfortunately, Cypress is not supported in Windows quite yet.

That's huge for me. Like, a potential ship-stopper :-( Many workshop attendees will be using windows machines. Yikes. I wish I had known about this before :-/ It may be that the most I can do for my workshop is mention Cypress and maybe show it off... It would be a poor experience for those on Windows machines otherwise.

Is there anything I can do to help make this supported on Windows by next month? cc @brian-mann

**Comment by brian-mann** [comment]
You can have them install VirtualBox Ubuntu and run it on there. That's what our current Windows users are doing.

Supporting Windows is... a big deal, we will need to hire at least 1 more full time person before we can support it. We've written all of our API's to support it, but it adds a lot of infrastructure complexity / support that we cannot possibly take on right now.

**Comment by kentcdodds** [comment]
Ok, that could work. Thanks @brian-mann. Do you have a tutorial for doing this anywhere? I definitely wont want to cover how to do it in my workshop, so I'll need good instructions for people on Windows, taking into account that some may have never used something like VirtualBox before.

**Comment by RandallKent** [comment]
We collaborated with the good folks at Xebia on a [VM](https://github.com/xebia/twc-workshop-vm) for the workshop at [Testworks Conf](http://testworksconf.com/) in October 2016.

Also, @fvanwijk (frequently in Gitter) might be able to chime in.

**Comment by kentcdodds** [comment]
Thanks for chiming in @RandallKent. @edm00se is actually working on making a Docker file and some documentation on how to set up the project with docker. I've never used 🐳 before, but I hear it's pretty awesome so if it works I think I'll go that route.

**Comment by RandallKent** [comment]
Very cool! Hopefully you all will share the finished product with the community 😄

**Comment by kentcdodds** [comment]
This repo is totally open source, so yeah, it'll be shared for sure. And if it's really good and easy to make generic, then we could add it to [`workshop-setup`](https://github.com/kentcdodds/workshop-setup) 😄

**Comment by fvanwijk** [comment]
Since there is no Windows support we went for an Ubuntu VM to have equal environment for everyone and to get up and running (almost) without any setup.

This is is the preparation for TWC 2016: http://testworksconf.com/preparation-guide
Later we did a meetup/workshop with a slightly modified VM: only Cypress workshop on VM, not all the TWC workshops.

We found out that the performance was rather bad, so installing Cypress yourself could be a better option. On OSX this is easy: just follow the [installation guide](https://docs.cypress.io/docs/installing-and-running), clone the [repo](https://github.com/cypress-io/cypress-example-todomvc/tree/workshop) and run `npm start`.
The kitchen sink repo is also cloned as reference for the assignments.

On Windows you still need the VM.

Your repo feels a bit over engineered, but I can be totally wrong because we only focused on trying out Cypress itself instead of managing a project including dependencies, unit/E2E tests, etc.

Please let me know if you have any specific questions. I am not a VM/Docker specialist, but if you want to go into that direction I will add some Xebians to this discussion.

**Comment by kentcdodds** [comment]
I think we're pretty well tested up. Thanks for all the help everyone!
